azplugins/SinusoidalChannelGeometry.h

```diff
@@ -0,0 +1,366 @@
+
+                }
+            else
+                {
+
+                Scalar delta = abs(0 - ((m_Amplitude*fast::cos(x0*m_pi_period_div_L)+ sign*m_h) - vel.z/vel.x*(x0 - pos.x) - pos.z));
+
+                Scalar n,n2;
+                Scalar s,c;
+
+                while( delta > target_presicion && counter < max_iteration)
+                    {
+                    fast::sincos(x0*m_pi_period_div_L,s,c);
+                    n  =  (m_Amplitude*c + sign*m_h) - vel.z/vel.x*(x0 - pos.x) - pos.z;  // f
+                    n2 = -m_pi_period_div_L*m_Amplitude*s - vel.z/vel.x;                  // df
+                    x0 = x0 - n/n2;                                                       // x = x - f/df
+                    delta = abs(0-((m_Amplitude*fast::cos(x0*m_pi_period_div_L)+sign*m_h) - vel.z/vel.x*(x0 - pos.x) - pos.z));
+                    counter +=1;
+                    }
+
+
+                /* The new z position is calculated from the wall equation to guarantee that the new particle positon is exactly at the wall
+                 * and not accidentally slightly inside of the wall because of nummerical presicion.
+                 */
+                z0 = (m_Amplitude*fast::cos(x0*m_pi_period_div_L)+sign*m_h);
+
+                /* The new y position can be calculated from the fact that the last position outside of the wall, the current position inside
+                 * of the  wall, and the new position exactly at the wall are on a straight line.
+                 */
+                y0 = -(pos.x-dt*vel.x - x0)*vel.y/vel.x + (pos.y-dt*vel.y);
+
+                // Newton's method sometimes failes to converge (close to saddle points, df'==0, bad initial guess,overshoot,..)
+                // catch all of them here and do bisection if Newthon's method didn't work
+                Scalar lower_x = fmin(pos.x - dt*vel.x,pos.x);
+                Scalar upper_x = fmax(pos.x - dt*vel.x,pos.x);
+
+                // found intersection is NOT in between old and new point, ie intersection is wrong/inaccurate.
+                // do bisection to find intersection - slower but more robust than Newton's method
+                if ( !(lower_x  <= x0 && x0 <= upper_x ))
+                    {
+                    Scalar3 point1 = pos;  //initial position
+                    Scalar3 point2 = pos-dt*vel; // final position at t+dt
+                    Scalar3 point3 = 0.5*(point1+point2); // halfway point
+                    Scalar fpoint1,fpoint3;
+                    //TODO: technically, the presicion of Newton's method and bisection is different. Should this be unified?
+                    while ((point1.x-point2.x) > target_presicion && counter < max_iteration)
+                        {
+                        fpoint1 = ((m_Amplitude*fast::cos(point1.x*m_pi_period_div_L)+ sign*m_h) - point1.z);
+                        fpoint3 = ((m_Amplitude*fast::cos(point3.x*m_pi_period_div_L)+ sign*m_h) - point3.z);
```

Comment:
Could you simplify this using the `isOutside` function? You could use that to bool test whether the point is inside/outside, then pick which points to use based on that.